'Canvas View': pfCanvas & pfCanvasEditor - Contribution from MiQ/SUI

dtaylor113 opened the pull request:
Ok, this is HUGE! -as in a lot of code :smile:   Six months work plus the original open source core lib code.

pfCanvasEditor
![image](https://cloud.githubusercontent.com/assets/12733153/22864386/786b82f6-f11e-11e6-93b2-d7ff90413778.png)

In Connection Mode:

![image](https://cloud.githubusercontent.com/assets/12733153/22864390/820a3514-f11e-11e6-8bc7-e7abed872183.png)

Some Notes:
- As stated in the ngDocs, this does NOT work in IE 11 or lower due to IE's lack of support for latest SVG specification.  Does work in FireFox, Chrome, and MS-Edge.
- pfCanvasEditor is a Angular 1.5 Component as it is the main/top component to use.  pfCanvas and other sub components are Angular 1.x Directives.
- There are several unit tests in MiQ/SUI, but they need to be modified to be used for pfCanvas & pfCanvasEditor.   I will create a JIRA story for the unit tests, as well as several other stories for follow up work for pfCanvas & pfCanvasEditor:
  [PTNFLY-2193](https://patternfly.atlassian.net/browse/PTNFLY-2193): Unit Tests
  [PTNFLY-2197](https://patternfly.atlassian.net/browse/PTNFLY-2197): Remaining Issues (contains links to the other Canvas View JIRA stories)

@serenamarie125, @bleathem, @catrobson

serenamarie125 commented:
@dtaylor113 thanks for finishing this off so quickly!  AFAIK, the remaining issues in PTNFLY-2197 are not needed.  @catrobson, can you confirm?

dtaylor113 commented:
@serenamarie125, @catrobson, ok on PTNFLY-2197, but just want to make sure you see the other 'Canvas View' stories which are linked to PTNFLY-2197:
[PTNFLY-2194](https://patternfly.atlassian.net/browse/PTNFLY-2194): Canvas View: Toolbox 'leaveCopy' functionality (AngularJS)
[PTNFLY-2195](https://patternfly.atlassian.net/browse/PTNFLY-2195): Canvas View: Panning functionality/cursor (AngularJS)
[PTNFLY-2196](https://patternfly.atlassian.net/browse/PTNFLY-2196): Canvas View: Left Toolbox (AngularJS)
Thanks,
- Dave

serenamarie125 commented:
@bleathem @jeff-phillips-18 @dlabrecq @dgutride is it possible to get some eyes on this?   This is something that we pushed to get @dtaylor113 to complete last week, since once of Cat's products is waiting for it.

jeff-phillips-18 commented on src/canvas-view/canvas-editor/canvas-editor.component.js:
```diff
@@ -0,0 +1,166 @@
+        newNodeCount++;
+        newNode.x = event.clientX - 600;
+        newNode.y = event.clientY - 200;
+        newNode.backgroundColor = newNode.backgroundColor ? newNode.backgroundColor : '#fff';
```
This should use pfUtils.colorPalette.white

dtaylor113 replied:
Weird, but there isn't a 'white' in patternfly.pfPaletteColors

jeff-phillips-18 replied:
Well that is silly...

jeff-phillips-18 commented on src/canvas-view/canvas-editor/canvas-editor.component.js:
```diff
@@ -0,0 +1,166 @@
+        newNodeCount++;
+        newNode.x = 250 + (newNodeCount * 4 + 160);
+        newNode.y = 200 + (newNodeCount * 4 + 160);
+        newNode.backgroundColor = newNode.backgroundColor ? newNode.backgroundColor : '#fff';
```
This should use pfUtils.colorPalette.white

jeff-phillips-18 commented on src/canvas-view/canvas-editor/canvas-editor.less:
```diff
@@ -0,0 +1,190 @@
+        }
+      }
+      .not-draggable {
+        background-color: #ebebeb;
```
Palette color?

jeff-phillips-18 commented on src/canvas-view/canvas/canvas-directive.js:
```diff
@@ -0,0 +1,440 @@
+  // Directive that generates the rendered chart from the data model.
+  //
+  .directive('pfCanvas', function ($document) {
+    return {
```
This should be a component.

dtaylor113 replied:
I agree, however it is not a requirement for this initial PR which requires pfCavnasView to be a component to be accessed in an Angular2 'hybrid mode'.  Converting pfCanvas to a component would allow Angular2 folks to access the basic canvas component (without toolbox & zoom) -which I'm not sure is a use case.   I've captured this in [PTNFLY-2197](https://patternfly.atlassian.net/browse/PTNFLY-2197): Remaining Issues

jeff-phillips-18 replied:
I believe the requirement for the 4.x release was to use components. We converted all the directives to components and don't think we should, at this point, introduce new directives.

dtaylor113 replied:
Again, I agree, but it will take a significant rewrite of that core directive in order to convert it to a component.  The main pfCanvasEditor is a component and is what end users can/will access from a hybrid angular 2 setup.    @dgutride and I talked about that child/sub directives used in component templates were ok (not ideal), as long as the parent was a component.

jeff-phillips-18 replied:
OK, I would have preferred not to introduce directives but I guess this could be refactored at a later time.

jeff-phillips-18 commented on src/canvas-view/canvas/canvas-viewmodel.js:
```diff
@@ -0,0 +1,1154 @@
+    //
+    this.backgroundColor = function() {
+      return this.data.backgroundColor || "#bbbbbb";
+    };
```
This should use pfUtils.colorPalette.black400

dtaylor113 replied:
canvas-viewmodel is the original open source canvas code and is defined as a global var which pfCanvas uses.  Unfortunately, canvas-viewmodel isn't compiled as part of the angular flow, so not sure how to inject pfUtils into it.   I could investigate $inject, but this use of #bbbbbb isn't really necessary since we default it to #ffff (see issue above :smile: ), so just removing it.

jeff-phillips-18 commented on src/canvas-view/canvas/canvas.less:
```diff
@@ -0,0 +1,240 @@
+.canvas-in-connection-mode {
+  background-image: none !important;
+  background-repeat: unset !important;
+  background-color: #b1b3b6;
```
Palette color.

jeff-phillips-18 commented on src/canvas-view/canvas/canvas.less:
```diff
@@ -0,0 +1,240 @@
+.invalid-node-header {
+  fill: #b1b3b6;
+  color: #b1b3b6;
+}
```
Palette colors

jeff-phillips-18 commented on src/canvas-view/canvas/canvas.less:
```diff
@@ -0,0 +1,240 @@
+  fill: transparent;
+  stroke-width: 2;
+  stroke: #0000ff;
+}
```
Palette color?

jeff-phillips-18 commented on src/canvas-view/canvas/canvas.less:
```diff
@@ -0,0 +1,240 @@
+  position: relative;
+  &::after {
+    border-bottom-color: @color-pf-white;
+    border-color: #333333;
```
Palette color

jeff-phillips-18 commented on src/canvas-view/examples/canvas.js:
```diff
@@ -0,0 +1,334 @@
+         width: 99%;
+     }
+   </style>
+   <div ng-controller="CanvasDemoCtrl" class="example-container">
```
These styles should go into the examples css

jeff-phillips-18 commented on src/canvas-view/examples/canvas.js:
```diff
@@ -0,0 +1,334 @@
+              "y": 67,
+              "id": 1,
+              "image": "/img/OpenShift-logo.svg",
+              "width": 150,
```
We should avoid using project related images and use generic images and names.

dtaylor113 replied:
These are included in core Patternfly dist/img.

jeff-phillips-18 replied:
Understood. My suggestion was to use generic items rather than storage specific items ie 'Dolor' rather than 'OpenShift'

jeff-phillips-18 commented:
![image](https://cloud.githubusercontent.com/assets/11633780/23067485/961463f4-f4ed-11e6-873f-f1be940f88b2.png)

Some alignment issues here:

Hide connectors is not centered on the toggle
Filter by name box should have some left padding before the text
Tabbed content should have some left padding from the tab area
The +/- buttons should have hover treatments
The duplicate and delete buttons should have hover treatments (when enabled)

jeff-phillips-18 commented:
Besides clicking on the Add Item again to close the add area, we should put a dismiss button in the upper right (x).

jeff-phillips-18 commented:
It is odd to me to have the filter below the items, everywhere else we have filtering above.

dtaylor113 commented:
Hi, addressed most issues raised by Jeff.  @serenamarie125 , please let me know your thoughts on Jeff's last two comments on the 'X' to close the toolbox and the placement of the toolbox filter.  Please let me know if you'd like me to work on these now, or add them to the JIRA story.  -thanks

jeff-phillips-18 commented on src/canvas-view/canvas-editor/canvas-editor.less:
```diff
@@ -39,7 +39,7 @@
     }
     .show-hide-connectors-label {
-      vertical-align: text-bottom;
+      vertical-align: 2px;
```
Removing the vertical-align setting all together seems to make it line up better.

dtaylor113 replied:
Not for me?  Seems to bottom align.
![image](https://cloud.githubusercontent.com/assets/12733153/23081767/5635afb8-f524-11e6-80b0-5acdc93186f2.png)

jeff-phillips-18 replied:
If I remove the setting I get:

![image](https://cloud.githubusercontent.com/assets/11633780/23081936/1ce4f2a4-f525-11e6-8710-89b5eec4bdf9.png)

jeff-phillips-18 commented on src/canvas-view/canvas-editor/canvas-editor.less:
```diff
@@ -0,0 +1,194 @@
+    position: relative;
+    width: 100%;
+    .canvas-editor-toolbox {
+      background-color: rgba(255, 255, 255, 0.94);
```
Do we want less than full opacity for this? It looks odd to me having the node items bleed through.

dtaylor113 replied:
Hmm..the original mockups specifically showed the items showing through the toolbox...

jeff-phillips-18 replied:
I'm looking at https://github.com/patternfly/patternfly-design/blob/master/pattern-library/content-views/canvas-view/design/design.md

I see the x to dismiss and a gray background not translucent.

dtaylor113 replied:
No, I'm talking way back to Colleen's original InVision mockups. I'm not sure the current design you mention above actually shows nodes behind the toolbox, and that they are hidden by the toolbox.

Either way is fine with me, easy enough to change the value :-)

dtaylor113 replied:
That design is rather recent, and is somewhat incomplete as it doesn't show the Filter in the Toolbox or the Hide Connections checkbox.

bleathem commented:
Thanks for the thorough review @jeff-phillips-18!

serenamarie125 commented:
@dtaylor113 those two remaining issues @jeff-phillips-18 requested sound to be low hanging fruit.  Are they things that could be knocked off quickly ?

dtaylor113 commented:
@serenamarie125, it was more a question of if you agree with the changes.  I can start work on them now.  Just don't know if this is holding up a release? ( @bleathem )

dtaylor113 commented:
Done.  New placements: 
![image](https://cloud.githubusercontent.com/assets/12733153/23082017/6f986ab2-f525-11e6-9d4e-22ad43c73bef.png)

serenamarie125 commented:
I meant the "x" to close the add and the vertical alignment of the hide connectors :) @dtaylor113

serenamarie125 commented:
I think the other things we can hold off.  Media Drive should be able to use these, and we can do a secondary review on this later on, imo.
